chain/rust/src/assets.rs

```diff
@@ -0,0 +1,595 @@
+                                Entry::Occupied(mut assets) => {
+                                    let current = assets.get_mut();
+                                    *current = current.checked_add(amount)?;
+                                }
+                                Entry::Vacant(vacant_entry) => {
+                                    vacant_entry.insert(*amount);
+                                }
+                            }
+                        }
+                        Entry::Vacant(entry) => {
+                            let mut assets = OrderedHashMap::new();
+                            assets.insert(asset_name.clone(), *amount);
+                            entry.insert(assets);
+                        }
+                    }
+                }
+            }
+        }
+        Some(Self(bundle))
+    }
+
+    /// Subtracts rhs from this bundle.
+    /// This does not modify self, and instead returns the result.
+    /// If underflow would occur on any asset, or the asset didn't exist in self
+    /// then None is returned.
+    /// Use clamped_sub (ClampedSub trait) if you need to only try to remove assets when they exist
+    /// and ignore them when they don't.
+    pub fn checked_sub(&self, rhs: &Self) -> Option<Self> {
+        let mut bundle = self.0.clone();
+        for (policy, rhs_assets) in rhs.iter() {
+            for (asset_name, rhs_amount) in rhs_assets.iter() {
+                match bundle.get_mut(policy) {
+                    Some(lhs_assets) => match lhs_assets.get_mut(asset_name) {
+                        Some(lhs_amount) => match lhs_amount.checked_sub(rhs_amount) {
+                            Some(new_lhs_amount) => {
+                                if new_lhs_amount.is_zero() {
+                                    lhs_assets.remove(asset_name);
+                                    if lhs_assets.is_empty() {
+                                        bundle.remove(policy);
+                                    }
+                                } else {
+                                    *lhs_amount = new_lhs_amount;
+                                }
+                            },
+                            None => {
+                                // underflow
+                                return None;
+                            }
+                        },
+                        None => {
+                            // asset name is missing from left hand side
+                            return None;
+                        }
+                    },
+                    None => {
+                        // policy id missing from left hand side
+                        return None;
+                    }
+                }
+            }
+        }
+        Some(Self(bundle))
+    }
+}
+
+pub type Mint = AssetBundle<i64>;
+
+pub type MultiAsset = AssetBundle<u64>;
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
```

Comment:
One other possibility, if we really want to maximize direct code generation, is to instead modify this API over top of 100% directly generated `value = coin / [coin, multiasset]`. That would result in something like:
```rust
pub enum Value {
    Coin {
        coin: Coin,
        #[serde(skip)]
        coin_encoding: Option<Sz>,
    },
    Multiasset(MultiAssetValue),
}

pub struct MultiAssetValue {
    coin: Coin,
    multiasset: MultiAsset,
    #[serde(skip)]
    encoding: Option<MultiAssetValueEncoding>,
}
```

I left it how it is (we generated this originally off of code like `value = [coin, multiasset]` then I just added 1 variable to `ValueEncoding` + spliced in the `Coin` part of (de)serialization) to closer match old CML's `Value` class and to simplify the multiasset handling for such a core class like `Value`.  Users can check `Value::has_multiassets()` if they want to check.

Reply:
i.e. then we'd have code like `assets/mod.rs` / `assets/cbor_encodings.rs` / `assets/serialization.rs` containing that code and then all this other stuff in `assets/utils.rs`.

Reply:
I think enum is not the most convenient solution for the `Value`. E.g. we can end up in a strange situation with `Multiasset` option, where `multiasset` field has no assets

Reply:
Yeah it's more of how far we want to take the idea of only writing custom code _on top of_ the generated code, rather than replacing it. And the only way to generate code directly here would be with the enum since it's a type choice.